internal/controller/controllers/local_cluster_import_controller.go

```diff
@@ -0,0 +1,711 @@
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/hashicorp/go-multierror"
+	configv1 "github.com/openshift/api/config/v1"
+	hiveext "github.com/openshift/assisted-service/api/hiveextension/v1beta1"
+	aiv1beta1 "github.com/openshift/assisted-service/api/v1beta1"
+	conditionsv1 "github.com/openshift/custom-resource-status/conditions/v1"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/apis/hive/v1/agent"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	agentServiceConfigLocalClusterImportFinalizerName = "agentserviceconfig." + aiv1beta1.Group + "/local-cluster-import-deprovision"
+)
+
+type LocalClusterImportReconciler struct {
+	Client                 client.Client
+	LocalClusterName       string
+	Log                    *logrus.Logger
+	AgentServiceConfigName string
+}
+
+func NewLocalClusterImportReconciler(client client.Client, localClusterName string, agentServiceConfigName string, log *logrus.Logger) *LocalClusterImportReconciler {
+	return &LocalClusterImportReconciler{
+		Client:                 client,
+		LocalClusterName:       localClusterName,
+		Log:                    log,
+		AgentServiceConfigName: agentServiceConfigName,
+	}
+}
+
+func (r *LocalClusterImportReconciler) setReconciliationStatus(agentServiceConfig *aiv1beta1.AgentServiceConfig, completed bool, reason string) error {
+	status := v1.ConditionFalse
+	if completed {
+		status = v1.ConditionTrue
+	}
+	conditionsv1.SetStatusConditionNoHeartbeat(&agentServiceConfig.Status.Conditions, conditionsv1.Condition{
+		Type:    aiv1beta1.ConditionReconcileCompleted,
+		Status:  status,
+		Reason:  aiv1beta1.ReasonReconcileSucceeded,
+		Message: reason,
+	})
+	err := r.Client.Update(context.Background(), agentServiceConfig)
+	if err != nil {
+		r.Log.Errorf("Unable to update status of ASC while attempting to set condition %s", aiv1beta1.ConditionReconcileCompleted)
+		return err
+	}
+	return nil
+}
+
+// +kubebuilder:rbac:groups=config.openshift.io,resources=dnses,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.open-cluster-management.io,resources=managedclusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=config.openshift.io,resources=proxies,verbs=get;list;watch
+
+func (r *LocalClusterImportReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var asc ASC
```

Comment:
I don't understand, this is so that we can use an appropriate go receiver to perform the reconciliation. Are you suggesting that we use the AgentServiceConfigReconciler instead?